draft-celi-wiggers-tls-authkem.md

```diff
@@ -177,17 +177,27 @@ server:  The endpoint that this did initiate the TLS connection.
 As this proposal relies heavily on KEMs, which are not originally
 used by TLS, we will provide a brief overview of this primitive.
 
-A Key Encapsulation Mechanism (KEM), defined as in {{!I-D.irtf-cfrg-hpke}}
-as an internal API, is a cryptographic primitive that defines the
-methods ``Encap`` and ``Decap``:
+A Key Encapsulation Mechanism (KEM) is a cryptographic primitive that defines
+the methods ``Encapsulate`` and ``Decapapsulate``:
 
-``Encap(pkR)``:  Takes a public key, and produces a shared secret and
+``Encapsulate(pkR)``:  Takes a public key, and produces a shared secret and
   encapsulation.
 
-``Decap(enc, skR)``:  Takes the encapsulation and the private key. Returns
+``Decapsulate(enc, skR)``:  Takes the encapsulation and the private key. Returns
   the shared secret.
 
-Note that we are using the internal API for KEMs as defined in {{!I-D.irtf-cfrg-hpke}}.
+We implement these methods through the KEMs defined in {{!I-D.irtf-cfrg-hpke}}
+as follows to export shared secrets appropriate for the use of HKDF in TLS 1.3:
+
+~~~
+def Encapsulate(pk):
```

Comment:
I might also pass in an info string here that's fed to the Export() API (for Encap/Decap separation when used multiple times in a single flow).

Reply:
I purposefully _didn't_ allow that, because otherwise you'll be stuck with having to decide on Kyber+which KDF all over again when we want to move to post-quantum KEMs, which seems unnecessary.

Reply:
Eh? I don't mean to suggest that the _algorithms_ be fed into the info parameter. Rather, when when `Encapsulate` is called on the server-side, it might be called with the "tls13 server" info parameter. (And likewise for the client.)

Reply:
I meant that it's not possible to pass "tls13 server" to Kyber512, so if we ever want to replace Encapsulate() by for example straight-up "Kyber.Encapsulate", you can't anymore; you'll always need HKDF to be able to still feed in this label.

Reply:
I... don't understand. You're not passing this to Kyber, you're passing it to HPKE, which takes care of that. If you're saying that Kyber can't fit underneath the HPKE API, that's a very different thing.

Reply:
Although I guess that if we want to keep using HPKE even in a post-quantum setting, then it's possible. I'm just not sure that extra complexity is warranted.

Reply:
My concern is that using the labels locks you into HPKE while the PQ KEM don't need the KDF in the same way that DH g^xy does (the KEMs already produce the result from a hash).

Reply:
The point of this change, to me, was specifically to introduce an HPKE dependency specifically to avoid yet another KEM algorithm registry and to build on the work that HPKE did to make this easy. "Standardized post-quantum" KEMs will find their way into HPKE's registry when available.

Reply:
> While the post-quantum KEMs are likely to be added to HPKE, don't you also expect them to be registered separately as "bare" algorithms?

I don't really see that happening. You'd not only need the algorithm registry, but then also define (yet another) way to implement domain separation. (It's true that the NIST KEMs don't need to KDF the output absent separation, but that doesn't seem to hold if you nee domain separation.)

Reply:
I agree with @thomwiggers on all. The PQ KEM don't need the usage of KDF in the same way that DH does. For me, this was a placeholder as well.

Reply:
What about domain separation? Are we punting on that entirely here?

Reply:
As pointed out elsewhere by Chris:
> From https://datatracker.ietf.org/doc/html/rfc8446#section-4.4.3:
The context string for a server signature is "TLS 1.3, server CertificateVerify".  The context string for a client signature is "TLS 1.3, client CertificateVerify". 

I don't think we considered this domain separation being present in the TLS CertificateVerify messages; this might be a good reason to include the domain separation feature of HPKE.